Allow regex in blacklisted dirs and files

tdaget opened the pull request:
## Description
I need to support regex in blacklist functionality for files and dirs. Need to set blacklist on the server (want to enforce corporate IT policies).
`excluded_directories` and `blacklisted_files` variables already exist in the parameter file but only work on exact folder name and filename.

In lib/private/Files/Filesystem.php : envolve `isFileBlacklisted` function in order to evaluate regex match for each entry declared in $excluded_directories and $blacklisted_files config parameters respectively for folder name and filename.
In config/config/sample.php : For $excluded_directories and $blacklisted_files variables, add chars to be regex compatible.

Regex syntax can be tested with [https://www.phpliveregex.com](https://www.phpliveregex.com)

## Motivation and Context
This change was initiated as in our structure several people sync there local Outlook pst file (despite IT recommendations). This generate a lot (a very lot) of file's change on local disk so made a lot of network traffic between Owncloud desktop client and server.

## How Has This Been Tested?
Currently, this patch is deployed on our test platform without any drawbacks. Community comments are welcome ...
Note: Regex match is CPU consuming, hope this will not raise major problem on a production platform. Stress tests must be performed.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:
- [x] Code changes
- [x] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [ ] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/owncloud/core?pullRequest=36331) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/owncloud/core?pullRequest=36331) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/owncloud/core?pullRequest=36331) it.</sub>

mmattel commented on lib/private/Files/Filesystem.php:
```diff
@@ -655,20 +655,26 @@ public static function isForbiddenFileOrDir($FileOrDir, $ed = []) {
+                if ($excluded) {
+                        $excluded = \array_map('trim', $excluded);
+                        $excluded = \array_map('strtolower', $excluded);
+                        foreach($path_parts as $path_part) {
```
* You could do a check here (and below) if `$excluded` contains a to be defined "forbidden" character (out of a list, for hints see: https://github.com/owncloud/core/issues/29759 plus usual regex characters, blacklist or whitelist what ever better fits) which would indicate a present regex and then branch between the old and your new regex querey. This would greatly improve performance as it continues using the old way with `array_intersect` and uses `preg_match` in the loop only if necessary.
* The query characters used for indicating a regex must be described in the config.sample file above.

* Add an example like:
```
hugo   // does not contain a regex character
will match the first item of the path component only:
hugo
hugo.test
testhugo

(?:hugo)  // contains regex
will match all items of the path component:
hugo
hugo.test
testhugo
```

tdaget replied:
Advise 2 and 3: done.
Advise 1: don't know how to implement them ...

mmattel commented on config/config.sample.php:
```diff
@@ -1219,8 +1221,8 @@
 	array (
-		'.snapshot',
-		'~snapshot',
+		'^\.snapshot$',
```
* I propose that you leave the original examples as written and add additional but different more fancy once to the list (blacklisted and excluded). Helps to make things more clear that you can use the old and the new (regex) way.
* In addition you must add a descriptive text that regex is allowed.
Add a `WARNING: ` message about the possible performance impact like:
```
WARNING: COMPLEX REGEX MAY SIGNIFICANT IMPACT YOUR SCANNING PERFORMANCE
```
The term warning + the colon + blank afterwards must be exacly written as in the example above because this has a doc rendering background.

tdaget replied:
Can't leave previous examples as they must be regex compliant (per example the dot in '.snapshot'). But more information and example might be welcome.
Warning added. Thanks

mmattel commented on lib/private/Files/Filesystem.php:
```diff
@@ -655,20 +655,26 @@ public static function isForbiddenFileOrDir($FileOrDir, $ed = []) {
+                        $excluded = \array_map('strtolower', $excluded);
+                        foreach($path_parts as $path_part) {
+                                foreach($excluded as $blackitem) {
+                                        if(preg_match('/'.$blackitem.'/i', $path_part)) {
```
Please use: `preg_match` --> `\preg_match` (and below)

See: https://doc.owncloud.com/server/10.3/developer_manual/core/code-standard.html

tdaget replied:
Thanks for the doc's reference, correction is done.

mmattel commented:
Your `Fixes` list from above is not correct as sync excludes are not addressed by this PR.
These are a different topic. Pls remove them

tdaget commented:
> Your `Fixes` list from above is not correct as sync excludes are not addressed by this PR.
> These are a different topic. Pls remove them

Done

mmattel commented on lib/private/Files/Filesystem.php:
```diff
@@ -658,16 +658,30 @@ public static function isForbiddenFileOrDir($FileOrDir, $ed = []) {
-				return true;
+			foreach($excluded as $blackitem) {
+				if(@\preg_match($blackitem, null) === false) {
+					\OC::$server->getLogger()->error('Malformed exclude regex: '.$blackitem.' - Check excluded_directories variable in config file.', ['app' => __CLASS__]);
```
1. Great that there is an error message now.
The message itelf only shows that something bad happened but gives no clue what it was.
It would be good to **add** a better hint about the cause. Please see the example below:
Based on:
https://www.php.net/manual/de/function.preg-last-error.php and https://stackoverflow.com/questions/4440626/how-can-i-validate-regex/12941133#12941133
I allowed myself to add/rename some stuff...
```
function is_preg_error()
{
    $errors = array(
        PREG_NO_ERROR               => 'Code 0 : No errors',
        PREG_INTERNAL_ERROR         => 'Code 1 : Internal PCRE error, most likely a syntax error',
        PREG_BACKTRACK_LIMIT_ERROR  => 'Code 2 : Backtrack limit was exhausted',
        PREG_RECURSION_LIMIT_ERROR  => 'Code 3 : Recursion limit was exhausted',
        PREG_BAD_UTF8_ERROR         => 'Code 4 : The offset did not correspond to the begin of a valid UTF-8 code point',
        PREG_BAD_UTF8_OFFSET_ERROR  => 'Code 5 : Malformed UTF-8 data',
        PREG_JIT_STACKLIMIT_ERROR   => 'Code 6 : Just-in-time compiler stack limit reached',
    );

    return $errors[\preg_last_error()];
}
```

tdaget replied:
Your code is added.
I put it directly in Filesystem.php class, don't know where to put it in a utility class ...

mmattel commented on lib/private/Files/Filesystem.php:
```diff
@@ -658,16 +658,30 @@ public static function isForbiddenFileOrDir($FileOrDir, $ed = []) {
+				} else {
+					foreach($path_parts as $path_part) {
+						if(\preg_match('/'.$blackitem.'/i', $path_part)) {
+							return true;
```
You may want to add a check if regex processing was ok:
`\preg_last_error() !== PREG_NO_ERROR`
and log an error in case using the function I described above.

tdaget replied:
Done

mmattel commented:
1. There are still coding style errors, pls see https://doc.owncloud.com/server/10.3/developer_manual/core/code-standard.html
and the build failing details from drone above

2. I am still not convinced hard switching from `array_intersect` done in one shot to `preg_match` double looping by default. I highly recommend as written in my comments above to at least do some checking for common regex characters and decide based on the result the way of checking. I know that there is no 100% identifyer, but I refrenced a hint in my comment. Things are easier for Windows backends as they have more excluding characters in filenames. But this is a matter of a proper description. This was the reason for keeping the old examples and adding new regex once.

tdaget commented:
@mmattel 
I agree, switching from `array_intersect` to `preg_match` might induce high CPU consumption.
I think we can consider following solutions:
1- create a new `excluded_directories` and a new `blacklisted_files` dedicated to regex match.
    I am scared of configuration difficulties and misunderstanding by users, I prefer to avoid this situation
2- allow a mix of strict filename and regex in these configuration variables, for each test is this is a string, if yes search for a direct match, if not try regex match. Double work ... And I don't know how to identify a simple string and invalid regex
3- consider that strict filename matching (previous way of working) is cover by regex use with an acceptable CPU cost
    The way I'm currently considering. Plus it permits to check regex validity systematically.

You propose to search for common regex characters, maybe not so reliable as these characters (for many of them and not least) can be used in filenames.

But all advice is welcome :-/

tdaget commented:
An idea to speed up the regex check: no need to check all `blacklisted_files` to all part of 'path to file' array, we need to match only the filename part.
Question (as I'm not an expert PHP developer): in the path_parts array variable, is the last item is ALWAYS (the always is very important here) the filename itself? In that case, we can update code with -drop foreach line 709, and replace `if (\preg_match('/'.$blackitem.'/i', $path_part)) {` by `if (\preg_match('/'.$blackitem.'/i', end($path_parts))) {` line 710.

mmattel commented:
Hope you are fine with all my comments... :smile:
1. You need to rebase as there was a change in the drone setup that needs to be populated into your PR so drone can process properly and turns into green 
2. `foreach ($excluded as $blackitem)` and `foreach ($blacklist as $blackitem)` :
only an optical thing, but to avoid confusion, please use a more neutral name for `$blackitem`
3. The code for checking `\preg_last_error()` should also run after each `foreach ($path_parts`.
This because you can run in all the other errors that syntax which is checked before. Recommending that the check and the error is put into a function which is called with an (optional) error text as you have written it already. This would reduce number of lines and makes the code more readble.
4. I get your point regarding regex / non regex confusion. To keep the positives of both variants, what about the idea to implement two new config parameters like `blacklisted_files_regex` and `excluded_directories_regex`? You then can run the code based on which one is set. Yes, you can set both but it gives admins the ability deciding which way is best for them. If both are set, you must run both and combine the result with `or`. Please a good descriptive text if you go for that approach.
5. Thanks for looking at `integrity.excluded.files`, but pls turn that back, as there is another PR already in review which takes care on that (and more).
6. Regarding `blacklisted_files`:
-Reminder, this code **must** always run, but not necessarily a empty regex version... (see idea above)
-The last element of `$path_parts` can either be the directoryname or the filename if a filename exists. You can improve the code as you thought and do the same for the current way if you want. Thanks for spotting that.

tdaget reviewed (commented):
Typo corrected in tests/data/app/invalid-info.xml
Add regular match for blacklisted folders and blacklisted files

mmattel commented on apps/federatedfilesharing/templates/settings-personal-general.php:
```diff
@@ -44,7 +44,7 @@ class='js-gs-share social-gnu'>
 			</button>
 			<button class="social-googleplus pop-up"
-				data-url='https://plus.google.com/share?url=<?php p(\urlencode($_['reference'])); ?>'/>
+				data-url='https://plus.google.com/share?url=<?php p(\urlencode($_['reference'])); ?>'>
```
Great for finding bugs / creating improvements.
As this one has nothing to do with regex, it must be put into a own PR.

tdaget replied:
Reverted

mmattel commented on config/config.sample.php:
```diff
@@ -1201,32 +1201,73 @@
+
+
+/**
+ * Blacklist specific directory names and disallow scanning, creating and renaming
```
`Blacklist specific directory` --> `Exclude specific directory`

tdaget replied:
Corrected

mmattel commented on config/config.sample.php:
```diff
@@ -1201,32 +1201,73 @@
+),
+
+/**
+ * Blacklist specific directory names based on regex expression. Scanning, creating and renaming
```
`Blacklist specific directory` --> `Exclude specific directory`

tdaget replied:
Corrected

mmattel commented on config/config.sample.php:
```diff
@@ -1201,32 +1201,73 @@
+),
+
+/**
+ * Exclude specific directory names and disallow scanning, creating and renaming
```
This block seems to be a leftover and forgotten to be deleted...

tdaget replied:
You're alright
Corrected

mmattel commented on config/config.sample.php:
```diff
@@ -1201,32 +1201,73 @@
  */
 'integrity.excluded.files' =>
-	array (
+	array(
```
As already stated, pls no change here as we already have an open PR which covers this issue (plus other stuff).

tdaget replied:
Can't see this point in my code ... 'll check after correcting all other points

mmattel replied:
pls re-check. still there...

mmattel commented on lib/private/Files/Filesystem.php:
```diff
@@ -625,28 +625,91 @@ public static function isIgnoredDir($dir) {
 	}
 
 	/**
-	* Check if the directory path / file name contains a Blacklisted or Excluded name
```
Pls keep this explanation text from the content point of view as it describes what is meant with each kind.
It is totally ok to move it down to line 688.

mmattel commented on settings/templates/help.php:
```diff
@@ -45,11 +45,12 @@
-			<?php p($l->t('Commercial support')); ?> ↗
-		</a>
-	</li>
+		<li>
```
As stated above, this has nothing to do with regex.
Pls create an own PR for fixing other issues than regex.

tdaget replied:
Reverted

mmattel commented on tests/data/app/invalid-info.xml:
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0"?>
-<info
+<info>
```
As stated above, this has nothing to do with regex.
Pls create an own PR for fixing other issues than regex.

tdaget replied:
Reverted

mmattel commented on lib/private/Files/Filesystem.php:
```diff
@@ -625,28 +625,91 @@ public static function isIgnoredDir($dir) {
+	}
+
+	/**
+	 * Manage regex blacklist against Folder of File.
```
Remove the word `blacklist`

tdaget replied:
Word dropped

mmattel commented on lib/private/Files/Filesystem.php:
```diff
@@ -625,28 +625,91 @@ public static function isIgnoredDir($dir) {
+	 * @param path Folder of File complete path contained in an array
+	 * @return boolean folder of file forbudden status
+	 */
+	private static function blacklistRegexAgainstFolderOrFile($blacklist, $path) {
```
Please use a neutral name instead of `$blacklist`

tdaget replied:
Corrected

mmattel commented on lib/private/Files/Filesystem.php:
```diff
@@ -625,28 +625,91 @@ public static function isIgnoredDir($dir) {
+	 * Manage regex blacklist against Folder of File.
+	 * @param blacklist Blacklist regex array as defined in Config file
+	 * @param path Folder of File complete path contained in an array
+	 * @return boolean folder of file forbudden status
```
`folder of file forbudden status` --> `folder of file regex status`

tdaget replied:
Changed

mmattel commented:
We are going into the right direction :smile:
I see that you are a big fan of the term blacklist...
Using it as a common name everywhere is a bad idea. The original naming was made by design.
- Have function and variable names based on the usecase which further reflects config.sample.
(means, use blacklist... where you run code for blacklist and excluded... where you run code for excluded)
- `blacklistRegexAgainstFolderOrFile`--> `checkRegexAgainstFolderOrFile`
- `regex_validy_check($regexToBeChecked)` --> `regex_validy_check($regexToBeChecked, $message="")`
Add a second parameter and handover the message via the caller (like `Check excluded_directories variable in config file`). This message differs either from blacklist or excluded. Rename the first part of the message `Exclude regex error:` --> `Regex error:`
- Based on the example `blacklist_folders` and `blacklist_folders_regex` (same for excluded), as stated above, you **must** run them if set and combine the results with `or` for a final return.

tdaget commented:
> We are going into the right direction 
> I see that you are a big fan of the term blacklist...
> Using it as a common name everywhere is a bad idea. The original naming was made by design.
> 
> * Have function and variable names based on the usecase which further reflects config.sample.
>   (means, use blacklist... where you run code for blacklist and excluded... where you run code for excluded)
> * `blacklistRegexAgainstFolderOrFile`--> `checkRegexAgainstFolderOrFile`
Changed

> * `regex_validy_check($regexToBeChecked)` --> `regex_validy_check($regexToBeChecked, $message="")`
>   Add a second parameter and handover the message via the caller (like `Check excluded_directories variable in config file`). This message differs either from blacklist or excluded. Rename the first part of the message `Exclude regex error:` --> `Regex error:`
Good idea
Corrected

> * Based on the example `blacklist_folders` and `blacklist_folders_regex` (same for excluded), as stated above, you **must** run them if set and combine the results with `or` for a final return.
Sorry, Can't understand this point.
As I modified the isForbiddenFileOrDir method the sequence of tests (exclude then exclude_regex then blacklist then blacklist_regex) do by construction an implicit OR ==> first test make a 'return true' and kill all other tests. And each test is done only if directive is set in config file.
In my understanding this permit to skip all non-mandatory work.
Why do you want to combine tests ? I'll make code less readable ...

> * Please rebase to make CI happy and green
I'm working on it, step after step, but lessons are hard to remember ...

mmattel commented on lib/private/Files/Filesystem.php:
```diff
@@ -625,28 +625,92 @@ public static function isIgnoredDir($dir) {
+
+	/**
+	* Check if the directory / file name is blacklisted in config.php.
+	* $ed and the query with can be redesigned if filesystem.php will get
```
Please use at least parts the original description (written in one of my comments already)
I have made a suggestion and added some words to blacklist to extend the usecase description
```
	/**
	* Check if the directory path / file name contains a Blacklisted or Excluded name
	* config.php parameter arrays can contain file names to be blacklisted or directory names to be excluded
	* Blacklist ... files that may harm the owncloud environment like a foreign .htaccess file or other unwanted files
	* Excluded  ... directories that are excluded from beeing further processed, like snapshot directories
	* $ed and the query with can be redesigned if filesystem.php will get a constructor where it is then possible 
	* to define the excluded directory names for unit tests
```

tdaget replied:
Sorry, dropped this block by mistake
Corrected

mmattel commented on lib/private/Files/Filesystem.php:
```diff
@@ -655,20 +719,32 @@ public static function isForbiddenFileOrDir($FileOrDir, $ed = []) {
 		}
+		if ($blacklist_files_regex) {
+			if (self::checkRegexAgainstFolderOrFile($blacklist_files_regex, \end($path_parts),
+													"Check blacklisted_files_regex variable in config file")) {
```
The intend is a bit too big :smile:

tdaget replied:
Corrected

mmattel commented on config/config.sample.php:
```diff
@@ -1217,23 +1234,38 @@
+
+/**
+ * Exclude specific directory names based on regex expression. Scanning, creating and renaming
+ * is disabled when folder match one of the regex.
```
Just put this into two lines for a better readbility:
```
 * Exclude specific directory names based on regex expression.
 * Scanning, creating and renaming is disabled when folder matches one of the regex.
```

tdaget replied:
Corrected

mmattel commented on config/config.sample.php:
```diff
@@ -1217,23 +1234,38 @@
+/**
+ * Exclude specific directory names based on regex expression. Scanning, creating and renaming
+ * is disabled when folder match one of the regex.
+ * Use of excluded_directories variable have precedence (evaluated before) on excluded_directories_regex variable.
```
`Use of...` --> `INFO: Use of ...`
The term info or warnig has a special meaning here.
It will be rendered in the documentation in a special optical way.

tdaget replied:
Added

mmattel commented:
Very good! only a view things to take care on :smile:

tdaget commented on lib/private/Files/Filesystem.php:
```diff
@@ -625,28 +625,92 @@ public static function isIgnoredDir($dir) {
+
+				foreach ($path as $path_part) {                       // foreach folder or file in given path
+					if (\preg_match('/'.$item.'/i', $path_part)       // regex match item
+							&& self::regexValidyCheck($item)) {       // check if regex error occur
```
Good idea, added.

tdaget commented on lib/private/Files/Filesystem.php:
```diff
@@ -655,20 +719,32 @@ public static function isForbiddenFileOrDir($FileOrDir, $ed = []) {
-			return true;
+		if ($exclude_folders_regex) {
+			if (self::checkRegexAgainstFolderOrFile($exclude_folders_regex, $path_parts,
+													"Check excluded_directories_regex variable in config file")) {
```
Corrected

tdaget commented on config/config.sample.php:
```diff
@@ -1201,12 +1201,29 @@
- * Blacklist a specific file or files and disallow the upload of files
- * with this name. `.htaccess` is blocked by default.
+ * Blacklist filenames based on regex expression.
+ * Use of blacklisted_files variable have precedence (evaluated before) on blacklisted_files_regex variable.
```
Added

tdaget commented on config/config.sample.php:
```diff
@@ -1201,12 +1201,29 @@
 /**
- * Blacklist a specific file or files and disallow the upload of files
- * with this name. `.htaccess` is blocked by default.
+ * Blacklist filenames based on regex expression.
```
Corrected

mmattel commented:
- `integrity.excluded.files` seems still not reverted, maybe just overseen.
- in config.sample, in the new regex examples, no need to put the array elements as remarks (ok for the description). 
- maybe a good idea to squash the commits now

mmattel commented on lib/private/Files/Filesystem.php:
```diff
@@ -625,28 +625,102 @@ public static function isIgnoredDir($dir) {
+	 * @return boolean
+	 */
+	private static function regexValidyCheck($regexToBeChecked, $message="") {
+		if (self::is_preg_error() !== PREG_NO_ERROR) {
```
Shouldnt this be `\preg_last_error()` as `is_preg_error` returns a string ?

tdaget replied:
Right

tdaget commented:
> * in config.sample, in the new regex examples, no need to put the array elements as remarks (ok for the description).

No need? Maybe but as sysadmin I know examples are always welcome, especially with regex

If you want absolutely to remove these lines of example maybe we can put them in comment block?

tdaget commented:
> * maybe a good idea to squash the commits now

Need some help for this (don't understand the "squash commits" ??)

mmattel commented:
I meant: `//    '.*\.pst$',     // di...` —> `    '.*\.pst$',     // di...`
It is removing the first two `//`

mmattel commented:
Example squashing the last three commits:
(https://stackoverflow.com/questions/5189560/squash-my-last-x-commits-together-using-git)
```
git reset --soft HEAD~3 &&
git commit
git push -f
````
At the moment you have 32 commits... (just see the number on top).
If you squash all 32 commits and push them, you see finally only one.
Note: the number of commits to squash must not exceed the existing number of commits of this PR. You do squashing when the work is close to final / or final already. This makes checking easier.

mmattel commented:
on top, pls set the checkmark that you have added unit tests

tdaget commented:
> on top, pls set the checkmark that you have added unit tests

Done

tdaget commented:
> I meant: `// '.*\.pst$', // di...` —> ` '.*\.pst$', // di...`
> It is removing the first two `//`

Remove the two `//` mean activating this regex by default (as everybody bases his config.php on the sample file). I'm afraid of confusion and questions for a lot of people.

At most I can consider doing this change if the `*.pst` regex as Outlook files raises a lot of undesirable network traffic and Owncloud activities.

tdaget committed:
Allow regex in blacklisted dirs and files

mmattel commented:
> activating this regex by default

No worry about that one. The name says **sample** and there is a description in the documentation not to just copying stuff but using it wisely... And only a limited set config options are used during installation. Anything else must be used by actively adding/configuring it to the running config file.

mmattel commented:
I just tried to help a bit and wanted to clone your branch but have seen that you did not used a branch.
Now this is somehow a bad thing because you will hardly come into sync with master and its ongoing changes without having a branch. But this is needed because of changes running CI. My proposal therefore would be the following:
- Create a branch from actual master and make/copy all the changes you did in this PR.
- Push it and make a new PR from that push.
- Close this PR

Big benefit: Clean, short and open for changes in master (easy rebaseable) and CI will turn green
If you cant do that, I am willing to help, but can do this earliest late evening today. Just let me know.

Great work!

tdaget committed:
Update config.sample.php

mmattel commented:
Closing in favour of https://github.com/owncloud/core/pull/36360.
I have improved your changes. Tested and runs like a charm :smile:
@tdaget please do any changes if necessary in #36360

tdaget commented:
> Closing in favour of #36360.
> I have improved your changes. Tested and runs like a charm 
> @tdaget please do any changes if necessary in #36360

Thanks for doing this.
I must admit that I was a little lost in the management of branches and PRs.
I'm still not comfortable with git / github ...

mmattel commented:
You could use smartgit (https://www.syntevo.com/smartgit/download/) available for various OS.
Makes life a lot easier when you start using git

tdaget commented:
Thanks for our advise.
I start to use github with the official 'github desktop' under linux
not so bad once we understand the basic principles ...

Thanks for your help in this PR
I've close all my branch/PR created in a desperate discover of tools. I created another branch named `Typo-corrections` in order to work on all small typo found during this PR.
